### ISSUE TITLE
`test_reopen` fails for forthcoming libhdf5-1.14.4

### ISSUE THREAD

**Issue by ajelenak** [h5py/h5py#2403]
Just fyi... These lines in the `test_reopen` of the test_file.py cause `OSError` with the code of the future libhdf5 1.14.4 release:
https://github.com/h5py/h5py/blob/078a481371dd6a46e260eeb196a4ef80e32c6a8a/h5py/tests/test_file.py#L926-L927
The change in behavior is likely caused by this (text from [RELEASE.txt](https://github.com/HDFGroup/hdf5/blob/8245d3ab75a3f897e92a8d3a885c3b3898ec1887/release_docs/RELEASE.txt#L687-L712)):
```
    - Fixed H5Fget_access_plist so that it returns the file locking
      settings for a file

      When H5Fget_access_plist (and the internal H5F_get_access_plist)
      is called on a file, the returned File Access Property List has
      the library's default file locking settings rather than any
      settings set for the file. This causes two problems:

        - Opening an HDF5 file through an external link using H5Gopen,
          H5Dopen, etc. with H5P_DEFAULT for the Dataset/Group/etc.
          Access Property List will cause the external file to be opened
          with the library's default file locking settings rather than
          inheriting them from the parent file. This can be surprising
          when a file is opened with file locking disabled, but its
          external files are opened with file locking enabled.

        - An application cannot make use of the H5Pset_elink_fapl
          function to match file locking settings between an external
          file and its parent file without knowing the correct setting
          ahead of time, as calling H5Fget_access_plist on the parent
          file will not return the correct settings.

      This has been fixed by copying a file's file locking settings
      into the newly-created File Access Property List in H5F_get_access_plist.

      This fix partially addresses GitHub issue #4011
```

**Comment by aragilar** [comment]
https://github.com/h5py/h5py/blob/066cad17312a592b26f59f91ef383dbde29bd142/h5py/_hl/files.py#L155-L156 and hence https://github.com/h5py/h5py/blob/066cad17312a592b26f59f91ef383dbde29bd142/h5py/h5p.pyx#L1518-L1525 are where this logic is handled. I'm not sure what `H5Pset_file_locking` is doing with `ignore_when_disabled=False`, as that's what's different to the other cases.

**Comment by ajelenak** [comment]
Should there be a note in the documentation about the change for the `best-effort` option? Someone may have used it and now it will throw an error.

**Comment by takluyver** [comment]
What does the error from this look like? Is it only when reopening an existing file, or does `best-effort` always fail now?

**Comment by ajelenak** [comment]
The error text is: `file locking 'ignore disabled locks' flag values don't match`.

> Is it only when reopening an already-open file

Yes. I reported in the context of this test.

**Comment by takluyver** [comment]
OK, so probably all that's needed is tweaking the test itself? I doubt reopening the same file with different options comes up often enough in the wild to be a big deal.

**Comment by takluyver** [comment]
Fixed by #2406, I believe.
